### PR TITLE
CI: Bump CI Lint base image from bionic to jammy

### PR DESCRIPTION
**Pull request by robigan**
As the title says, this PR aims to address https://github.com/bitcoin/bitcoin/issues/26548 by switching the lint CI job from ubuntu's bionic base image which has python 3.6.5 to ubuntu jammy and use pyenv to install the correct python version as dictated in the project's .python-version file.

I have tested this in a VS Code Dev Container to as closely emulate the Cirrus CI Service

### CODE REVIEW

**Review by maflcko — COMMENTED** [review]
looks like the CI is aborting green with empty output?

https://cirrus-ci.com/task/6096893982277632?logs=lint#L1575

File: ci/lint/04_install.sh [hunk 1]
```diff
@@ -7,14 +7,19 @@
...
-  # Temporary workaround for https://github.com/bitcoin/bitcoin/pull/26130#issuecomment-1260499544
-  # Can be removed once the underlying image is bumped to something that includes git2.34 or later
-  sed -i -e 's/bionic/jammy/g' /etc/apt/sources.list
-  ${CI_RETRY_EXE} apt-get update
-  ${CI_RETRY_EXE} apt-get install -y --reinstall git
-)
+${CI_RETRY_EXE} apt-get install -y curl git gawk jq
+${CI_RETRY_EXE} apt-get install -y make build-essential libssl-dev zlib1g-dev \
+libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm \
+libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev
```
[the remark this answers is not in the document]
Comment (robigan) [reply ci/lint/04_install.sh:13]: Yes just noticed, gonna inspect this rq to see why this happens. As for the deps, pyenv requires this according to the docs

[the remark this answers is not in the document]
Comment (robigan) [reply ci/lint/04_install.sh:13]: Pyenv seems to exit prematurely and somehow I stupidly didn't pick this up.

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply ci/lint/04_install.sh:13]: Ah interesting. I wonder if this even works then. Initially, I tried using https://launchpad.net/%7Edeadsnakes/+archive/ubuntu/ppa , however it notes " for jammy, older python versions requre libssl<3 so they are not currently built".

**Comment by robigan** [comment]
@MarcoFalke I have addressed the issues in the new commit, and not follows through with the rest of the testing. Except that the test scripts in the repo seem to error out with git (See attached)
image
I am not going to try to address this issue as it's got to do with the linting scripts themselves which are interacting with Git, and Git throwing an issue. And given that I haven't touched any of that I'll leave it be unless I am missing something which in that case I'll try to remedy the issue that this change causes

File: ci/lint/04_install.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (robigan) [reply ci/lint/04_install.sh:13]: In my dev container, it does indeed build the python version without issue. I of course used the ubuntu:jammy image

[the remark this answers is not in the document]
Comment (robigan) [reply ci/lint/04_install.sh:13]: > Why is all of this needed? Is pyenv compiling itself and then the python version?

If this was Github Actions I would have cached the build of Python 3.6.15, but I don't think Cirrus supports this

**Comment by MarcoFalke** [comment]
Looks like it does OOM

**Comment by robigan** [comment]
Just took a look at Deadsnake's PPA. They claim they don't build Python 3.6.15 because it's already built by ubuntu's repos. Yet the files list contains a tar ball for python 3.6.15 for focal fossa release. I'd say it's possible to untar the tar ball into pyenv's version if it all works smoothly. As for ssl, I am not sure, how come does the test ci job use jammy despite the repo's python version being set to python 3.6.15?

**Comment by MarcoFalke** [comment]
I'd guess that maybe python 3.6 supports building on Jammy by now? You may be able to fix the OOM by bumping the memory or by using `clang`. Though, I think we'll also want to cache the python binary, otherwise too much time is spent on compiling it for every push.

**Comment by robigan** [comment]
> I'd guess that maybe python 3.6 supports building on Jammy by now? You may be able to fix the OOM by bumping the memory or by using `clang`. Though, I think we'll also want to cache the python binary, otherwise too much time is spent on compiling it for every push.

Python 3.6.15 with pyenv definitely compiles (at least for me which was in a dev container which is pretty close to a real ubuntu jammy based machine). The issue is caching the builds. I'll try bump the memory and see if that works, I don't think using Clang is possible as it doesn't seem that pyenv supports it, and using a custom script to download python and compile it requires maintenance that I can't assure. Using pyenv makes maintaining way easier and simple.

**Comment by robigan** [comment]
I have discovered that it is indeed possible to cache builds. I have added more commits that try to implement this as automagically as possible

**Comment by robigan** [comment]
@MarcoFalke The fixes have been implemented. I just don't understand why lint fails despite saying that no issues were found

**Comment by robigan** [comment]
It actually works. Finally. @MarcoFalke Sorry for the excessive amount of pings, but feel free to merge this. Lmk if any squashing is necessary

**Comment by hebasto** [comment]
> Lmk if any squashing is necessary

Yes, it is necessary.

**Comment by robigan** [comment]
> > Lmk if any squashing is necessary
> 
> Yes, it is necessary.

Now that I have done a destructive action, I start to believe that I can't have squashed those commits in the first place...

**Comment by robigan** [comment]
Now that I've slept on it, I think it might convene me to just generate a patch file for the files I have modified, and open up a new PR. As I forgot merging instead of rebasing does this kind of stuff

**Comment by robigan** [comment]
Yea no I'll do that

**Pull request closed without merging**